packages/controllers/src/permissions/Caveat.ts

```diff
@@ -0,0 +1,257 @@
+  /**
+   * Any additional data necessary to enforce the caveat.
+   *
+   * TODO:TS4.4 Make optional
+   */
+  readonly value: Value;
+};
+
+/**
+ * The {@link Caveat} factory function. Naively constructs a new caveat from the
+ * inputs. Sets `value` to `null` if no value is provided.
+ *
+ * @param type - The type of the caveat.
+ * @param value - The value associated with the caveat, if any.
+ * @returns The new caveat object.
+ */
+export function constructCaveat<Type extends string, Value extends Json>(
+  type: Type,
+  value: Value,
+): Caveat<Type, Value> {
+  return { type, value };
+}
+
+// Next, we define types used for specifying caveats at the consumer layer,
+// and a function for applying caveats to a restricted method request. This is
+// Accomplished by decorating the restricted method implementation with the
+// the corresponding caveat functions.
+
+/**
+ * A function for applying caveats to a restricted method request.
+ *
+ * @template ParentCaveat - The caveat type associated with this decorator.
+ * @param decorated - The restricted method implementation to be decorated.
+ * The method may have already been decorated with other caveats.
+ * @param caveat - The caveat object.
+ * @returns The decorate restricted method implementation.
+ */
+export type CaveatDecorator<ParentCaveat extends CaveatConstraint> = (
+  decorated: AsyncRestrictedMethod<RestrictedMethodParameters, Json>,
+  caveat: ParentCaveat,
+) => AsyncRestrictedMethod<RestrictedMethodParameters, Json>;
+
+/**
+ * TODO:docs
+ */
+type ExtractCaveatValueFromDecorator<Decorator extends CaveatDecorator<any>> =
+  Decorator extends (
+    decorated: any,
+    caveat: infer ParentCaveat,
+  ) => AsyncRestrictedMethod<any, any>
+    ? ParentCaveat extends CaveatConstraint
+      ? ParentCaveat['value']
+      : never
+    : never;
+
+/**
+ * TODO:docs
+ */
+export type CaveatValidator<ParentCaveat extends CaveatConstraint> = (
+  caveat: { type: ParentCaveat['type']; value: unknown },
+  origin?: string,
+  target?: string,
+) => void;
+
+/**
+ * TODO:docs
+ */
+export type CaveatSpecificationConstraint = {
+  /**
+   * The string type of the caveat.
+   */
+  type: string;
+
+  /**
+   * The decorator function used to apply the caveat to restricted method
+   * requests.
+   */
+  decorator: CaveatDecorator<any>;
+
+  /**
+   * The validator function used to validate caveats of the associated type
+   * whenever they are instantiated. Caveat are instantiated whenever they are
+   * created or mutated.
+   *
+   * The validator should throw an appropriate JSON-RPC error if validation fails.
+   *
+   * If no validator is specified, no validation of caveat values will be
+   * performed. Although caveats can also be validated by permission validators,
+   * validating caveat values separately is strongly recommended.
+   */
+  validator?: CaveatValidator<any>;
+};
+
+/**
+ * TODO:docs
+ */
+export type CaveatSpecification<SpecifiedCaveat extends CaveatConstraint> = {
```

Comment:
Not quite, it is used in `CaveatSpecificationMap` to enforce the mapping between the map keys and the `type` of each caveat.

Reply:
Oh I see, I misread the `Caveat<Key, any>` as `Caveat<any, any>`.

I still don't think it's doing anything though. It looks equivalent to

```
{
  [Key in Specification['type']]: Specification;
}
```

The condition is verifying that `Specification['type']` matches itself. It always matches itself.

Reply:
Hm, this may have done something when I did more validation but the effect appears to be the same no matter what I do at this point. Went for an even simpler `Record` solution in 14ec184